onnxruntime/core/providers/cuda/nn/conv.h

```diff
@@ -40,13 +88,16 @@ struct CudnnConvState {
   // these would be recomputed if x/w dims change
   std::vector<int64_t> y_dims;
   size_t workspace_bytes;
-  AlgoType algo;
+  decltype(AlgoPerfType().algo) algo;
   CudnnTensor x_tensor;
   CudnnFilterDescriptor filter_desc;
   CudnnTensor b_tensor;
   CudnnTensor y_tensor;
   CudnnConvolutionDescriptor conv_desc;
 
+  using input_shapes = std::pair<std::vector<int64_t>, std::vector<int64_t>>;
+  lru_map<input_shapes, AlgoPerfType> cached_benchmark_results { 10000 };
```

Comment:
Yes, I also don't like that. Two questions: 
1) Would it make sense to exclude the filter size changes? It seems to me that they should almost never change, right? So, we would cache only (input shape -> algo) and when the filter shape changes, we would discard all cached results and start over. That would make things a bit easier.
2) I also see no other way than hashing, but I'm a bit worried about collisions. So I would probably define a hash for `vector<int64_t>` to use in the `unordered_map` but still keep `vector<int64_t>` as the key so that collisions don't result in the wrong algo being used.
I'll implement this and the other changes tomorrow, thanks for the comments.

Reply:
Yes filter size won't change, and we can remove it from key. The idea of using hash is to avoid cost of log(n) comparison in searching a tree-based map implementation, where n could be as big as 10000. To avoid collison you may use unordered_multimap and store the actual shape in the value for a full comparison.

Reply:
OK, I'll remove the filter shape.

I understand the need for `unordered_map`. But if I would use the hash as key type and handle collisison with an `unordered_multimap` it's much simpler to just use `unordered_map<Key, T, MyHashFunctorForVector<int64_t>>` as I suggested above.